### DISCUSSION TITLE
Expression for energy gradient in hf/dft

### DISCUSSION THREAD

**Discussion by computative** [pyscf/pyscf#2656] [Q&A] [answered]
Hello there,

I'm trying to read the code to determine which analytical expression is used for the energy gradients for dft. I have the impression that the bulk of the work happens in [GradientsBase](https://pyscf.org/_modules/pyscf/grad/dhf.html#GradientsBase). But I don't know all the methods well enough to work out the formulas that have been programmed (in reasonable time; say in less than a few days).

So it would be helpful if you could provide the equation that is being used to calculate the forces for dft. I'd like to use pyscf for my project, but I need to understand how the forces are calculated.

Thanks for reading

**Comment by fishjojo** [comment, marked as answer]
you may refer to this paper https://doi.org/10.1063/1.464906

**Comment by computative** [comment]
Thanks for your message. Is it equation 9 that you have programmed in there for the dft energy gradient?

**Comment by sunqm** [comment]
The program implementation does not strictly follow a single literature source. It has been extensively optimized based on a slower initial implementation, which you can find https://github.com/pyscf/pyscf/blob/3ca5a34288fca6217177a094000cfc1eb51dd67f/pyscf/grad/test/test_tdrhf_grad.py#L31. You can refer to the paper to understand the equations, and then follow the slow implementation to see how it is translated into the PySCF code.

**Comment by verena-neufeld** [comment]
Hi @computative , thanks for your question. Has it been answered and can this issue be closed?

**Comment by computative** [comment]
yes

På onsdag 15. januar 2025 kl. 23:52, Verena Neufeld ***@***.***> skrev:
